sklearn/utils/tests/test_extmath.py

```diff
@@ -161,6 +162,85 @@ def test_randomized_svd_low_rank_all_dtypes(dtype):
     check_randomized_svd_low_rank(dtype)
 
 
+@pytest.mark.parametrize('dtype',
+                         (np.int32, np.int64, np.float32, np.float64))
+def test_randomized_eigsh(dtype):
+    """Test that `_randomized_eigsh` returns the appropriate components"""
+
+    X = np.diag(np.array([1., -2., 0., 3.], dtype=dtype))
+
+    # with 'module' selection method, the negative eigenvalue shows up
+    lambd_, alph_ = _randomized_eigsh(X, n_components=2, selection='module')
+    # eigenvalues
+    assert lambd_.shape == (2,)
+    assert_array_almost_equal(lambd_, [3., -2.])  # negative eigenvalue here
+    # eigenvectors
+    assert alph_.shape == (4, 2)
+
+    # with 'value' selection method, the negative eigenvalue does not show up
+    with pytest.raises(NotImplementedError):
+        _randomized_eigsh(X, n_components=2, selection='value')
+
+
+@pytest.mark.parametrize('k', (10, 50, 100, 200))
+def test_randomized_eigsh_compared_to_others(k):
+    """Check that `_randomized_eigsh` is similar to other `eigsh`
+
+    Tests that for a random PSD matrix, `_randomized_eigsh` provides results
+    comparable to LAPACK (scipy.linalg.eigh) and ARPACK
+    (scipy.sparse.linalg.eigsh)
+    """
+
+    # make a random PSD matrix
+    n_features = 200
+    X = make_sparse_spd_matrix(n_features, random_state=0)
+
+    # compare two versions of randomized
+    # rough and fast
+    lambdas, alphas = _randomized_eigsh(X, n_components=k, selection='module',
+                                        random_state=0)
+    # more accurate but slow (TODO find realistic settings here)
+    lambdasQr, alphasQr = _randomized_eigsh(X, n_components=k, n_iter=25,
+                                            n_oversamples=20,
+                                            power_iteration_normalizer="QR",
+                                            selection='module', random_state=0)
+
+    # with LAPACK
+    lambdas_lapack, alphas_lapack = linalg.eigh(X, eigvals=(n_features - k,
+                                                            n_features - 1))
+    indices = lambdas_lapack.argsort()[::-1]
+    lambdas_lapack = lambdas_lapack[indices]
+    alphas_lapack = alphas_lapack[:, indices]
+
+    # and ARPACK
+    v0 = _init_arpack_v0(n_features, random_state=0)
+    # Note: "LA" largest algebraic <=> selection="value" in randomized_eigsh
+    lambdas_arpack, alphas_arpack = eigsh(X, k, which="LA", tol=0,
+                                          maxiter=None, v0=v0)
+    indices = lambdas_arpack.argsort()[::-1]
+    lambdas_arpack = lambdas_arpack[indices]
+    alphas_arpack = alphas_arpack[:, indices]
+
+    # -- eigenvalues comparison
+    assert lambdas_lapack.shape == (k,)
+    assert_array_almost_equal(lambdas_lapack, lambdas_arpack, decimal=10)
+    # quite poor comparison precision. Is this related to the shape of the
+    # spectrum in the generated random dataset ?
+    assert_array_almost_equal(lambdas, lambdas_lapack, decimal=1)
```

Comment:
Maybe try to increase `n_oversamples` or the number of power iterations?

Reply:
Indeed increasing `n_iter` did the trick. It is surprising that `n_iter=10` is the default with such a difference in precision...